### PR TITLE
Add FV skewness-correction to Navier-Stokes module

### PR DESCRIPTION
**Pull request by grmnptr**
Refs. #16239

## Reason
To enable skewness-correction of the variables used in the finite volume Navier-Stokes objects.

## Design
Propagate changes of face-interpolation and gradient evaluation from MooseVariableFV.

## Impact
Might increase the accuracy/robustness of the finite volume Navier-Stokes solvers.

### CODE REVIEW

File: modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C [hunk 1]
```diff
@@ -529,10 +539,21 @@ INSFVMomentumAdvection::interpolate(Moose::FV::InterpMethod m, ADRealVectorValue
...
     neighbor_D(i) = neighbor_volume / neighbor_a(i);
   }
+
+  // We require this to ensure that the correct interpolation weights are used.
+  // This will change once the traditional weights are replaced by the weights
+  // that are used by the skewness-correction.
+  Moose::FV::InterpMethod face_velocity_interp_method =
+      (_advected_interp_method == Moose::FV::InterpMethod::SkewCorrectedAverage)
+          ? _advected_interp_method
+          : Moose::FV::InterpMethod::Average;
```
Comment (lindsayad) [inline modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C:549]: Hmm it seems weird to have this tied to the advected interpolation method, but since the velocity interpolation method should always be Rhie-Chow, I think this should be fine for now. If we decide that we will in fact always do Rhie-Chow (which we probably should decide that), then perhaps we can change `velocity_interp_method` to have the logical meaning of "how do you want to interpolate the velocity before performing the RC correction to it"

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i [hunk 2]
```diff
@@ -1,35 +1,32 @@
-mu=1.1
-rho=1.1
+mu=1.0
+rho=1.0
```
Comment (lindsayad) [inline modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i:2]: Why this change? I often do these settings just to make sure that I didn't actually forget a rho or mu coefficient somewhere.

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i [hunk 3]
```diff
@@ -1,35 +1,32 @@
...
+  error_on_jacobian_nonzero_reallocation=true
 []
 
-[Problem]
-  coord_type = 'RZ'
+[Mesh]
+  [./gen_mesh]
+    type = FileMeshGenerator
+    file = skewed.msh
+  [../]
```
Comment (lindsayad) [inline modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i:13]: Prefer `[]` over `[./]` and `[../]`

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i [hunk 4]
```diff
@@ -1,35 +1,32 @@
...
-    xmin = 1
-    xmax = 3
-    ymin = -1
-    ymax = 1
-    nx = 2
-    ny = 2
-  []
+[Problem]
+  coord_type = 'XYZ'
+  error_on_jacobian_nonzero_reallocation=true
```
Comment (lindsayad) [inline modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i:6]:
```suggestion
  error_on_jacobian_nonzero_reallocation = true
```

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i [hunk 5]
```diff
@@ -126,79 +118,62 @@ rho=1.1
...
-    type = ParsedFunction
-    value = 'sin(y)'
-    []
-  [forcing_p]
-    type = ParsedFunction
-    value = '-1/2*pi*rho*sin(x)*sin((1/2)*y*pi) + (x*pi*rho*sin(y)*cos(x*pi) + rho*sin(y)*sin(x*pi))/x'
-    vars = 'rho'
-    vals = '${rho}'
-  []
+[exact_u]
```
Comment (lindsayad) [inline modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i:143]: indent all these sub-blocks

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/test.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+        fig = mms.ConvergencePlot(xlabel='Element Size ($h$)', ylabel='$L_2$ Error')
+        fig.plot(df1, label=labels, marker='o', markersize=8, num_fitted_points=3, slope_precision=1)
+        fig.save('skewed.png')
+        for key,value in fig.label_to_slope.items():
+            print("%s, %f" % (key, value))
+            if key in velocity_labels:
+                self.assertTrue(fuzzyAbsoluteEqual(value, 2., .15))
+            else:
+                self.assertTrue(fuzzyAbsoluteEqual(value, 1., .2))
```
Comment (lindsayad) [inline modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/test.py:28]: Do you know why we lose an order for pressure?

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/tests [hunk 7]
```diff
@@ -1,11 +1,15 @@
...
+    input = test.py
+    test_case = TestVortexSkewCorrected
+    method = '!dbg'
+    min_parallel = 2
+    requirement = 'The system shall be able to solve the incompressible Navier-Stokes equations on triangular meshes, using a Rhie-Chow scheme and skewness-correction, Dirichlet boundary conditions for the velocity, and demonstrate second order convergence for the velocity and first order convergence for pressure.'
     ad_indexing_type = 'global'
-    requirement = "The system shall trow an error if the cell gradient is computed with skewness-correction in the INSFVVelocity variable."
-    expect_err = "INSFVelocityVariable does not support skewness-correction at this point!"
+    required_python_packages = 'pandas matplotlib'
+    heavy = true
```
Comment (lindsayad) [inline modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/tests:13]: Does this actually have to be marked `heavy`? How long does this test take?

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/tests (hunk 7 above)
Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/tests:13]: The whole mms case takes about 50 s on my local machine.

**Comment by lindsayad** [comment]
When this is merged, should we consider #16239 closed?

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/tests (hunk 7 above)
Comment (lindsayad) [reply modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/tests:13]: What if you use like 8 procs?

Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/tests:13]: With 8 procs it is 11 s.

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i (hunk 2 above)
Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i:2]: I set those to 1.0 because the formulation in the paper I used didn't bother adding the material properties in their formulation. The analytic solutions are only valid for this setting.

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/tests (hunk 7 above)
Comment (lindsayad) [reply modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/tests:13]: Ok I think I'm fine with 2 and running heavy or going with 8 and running non-heavy. Up to you

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i (hunk 2 above)
Comment (lindsayad) [reply modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i:2]: got ya

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/tests (hunk 7 above)
Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/tests:13]: I'll keep it heavy with 2.

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i (hunk 3 above)
Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i:13]: Got it! The next commit will fix this.

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i (hunk 5 above)
Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/skewed-vortex.i:143]: Got it! The next commit will fix this.

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/test.py (hunk 6 above)
Comment (grmnptr) [reply modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/test.py:28]: I have no idea, I haven't found good papers about this. It might have something to do with the Rhie-Chow interpolation. I like to think about it gaining 0.5 order of pressure :P.

File: modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C (hunk 1 above)
Comment (grmnptr) [reply modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C:549]: I agree, I think the Rhie-Chow interpolation falls into a different category. It is one layer above the average/vertex-based/skew-corrected. I'd like to work next year on a system which organizes these interpolation/limiter branches, I think the proposed modification can be a part of that.

File: framework/src/fvkernels/FVAdvection.C [hunk 8]
```diff
@@ -18,12 +18,13 @@ FVAdvection::validParams()
...
 
-  params.addParam<MooseEnum>("advected_interp_method",
-                             advected_interp_method,
-                             "The interpolation to use for the advected quantity. Options are "
-                             "'upwind' and 'average', with the default being 'upwind'.");
+  params.addParam<MooseEnum>(
+      "advected_interp_method",
+      advected_interp_method,
+      "The interpolation to use for the advected quantity. Options are "
+      "'upwind', 'average' and 'skewness-corrected', with the default being 'upwind'.");
```
Comment (GiudGiud) [inline framework/src/fvkernels/FVAdvection.C:27]: we dont need to list the options in the param description, moosedocs does it already

but it feels like skewness-correction should be a boolean to the kernels, even though for us it's a different interpolation method. 

when we ll have more options than average and upwind there, we ll want a boolean on skewness correction

File: framework/src/variables/MooseVariableFV.C [hunk 9]
```diff
@@ -1045,7 +1045,7 @@ MooseVariableFV<OutputType>::adGradSln(const Elem * const elem, const bool corre
                 "I believe we should only get singular systems when two-term boundary expansion is "
                 "being used");
     const_cast<MooseVariableFV<OutputType> *>(this)->_two_term_boundary_expansion = false;
-    const auto & grad = adGradSln(elem);
+    const auto & grad = adGradSln(elem, correct_skewness);
```
Comment (GiudGiud) [inline framework/src/variables/MooseVariableFV.C:1048]: umm since you added this one, do we also need this in evaluateGradient and in getInternalFaceValue?

File: modules/navier_stokes/doc/content/modules/navier_stokes/insfv.md [hunk 10]
```diff
@@ -75,3 +75,12 @@ line. An axisymmetric example with free slip conditions, using the Rhie-Chow
 interpolation is shown below:
 
 !listing modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/cylindrical/2d-rc-slip.i
+
+## Skewness-correction
+
+The skewness-correction of different variables can be enabled by defining the
+`face_interp_method=skewness-corrected` parameter for the INSFVVaribles and
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/modules/navier_stokes/insfv.md:82]:
```suggestion
`face_interp_method=skewness-corrected` parameter for the INSFVVariables and
```

File: modules/navier_stokes/doc/content/modules/navier_stokes/insfv.md [hunk 11]
```diff
@@ -75,3 +75,12 @@ line. An axisymmetric example with free slip conditions, using the Rhie-Chow
...
+selecting it as an option in the advection kernels. It has proven to increase
+accuracy on unstructured grids. For an example see:
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/modules/navier_stokes/insfv.md:84]: Can you say how much it increased? Is the conclusion the same across the board?

File: modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C [hunk 12]
```diff
@@ -408,23 +408,24 @@ INSFVMomentumAdvection::coeffCalculator(const Elem & elem) const
     mooseAssert((neighbor == &fi->elem()) || (neighbor == fi->neighborPtr()),
                 "Surely the neighbor has to match one of the face information's elements, right?");
 
-    ADRealVectorValue neighbor_velocity(_u_var->getNeighborValue(neighbor, *fi, elem_velocity(0)));
+    // Checking if skewness correction is necesary
```
Comment (GiudGiud) [inline modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C:411]:
```suggestion
    // Checking if skewness correction is necessary
```

File: modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C [hunk 13]
```diff
@@ -408,23 +408,24 @@ INSFVMomentumAdvection::coeffCalculator(const Elem & elem) const
...
+    bool correct_skewness =
+        (_u_var->faceInterpolationMethod() == Moose::FV::InterpMethod::SkewCorrectedAverage);
+
+    // Fill a face velocity for the advection contribution
+    ADRealVectorValue face_velocity(
+        _u_var->getInternalFaceValue(neighbor, *fi, elem_velocity(0), correct_skewness));
     if (_v_var)
```
Comment (GiudGiud) [inline modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C:418]: should we check that u, v and w all request skewness correction ?
I reckon in some meshes (extruded) it could make sense to skew-correct only some directions but feels pretty niche

File: modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C [hunk 14]
```diff
@@ -529,10 +539,21 @@ INSFVMomentumAdvection::interpolate(Moose::FV::InterpMethod m, ADRealVectorValue
...
+          ? _advected_interp_method
+          : Moose::FV::InterpMethod::Average;
+
   Moose::FV::interpolate(
-      Moose::FV::InterpMethod::Average, face_D, elem_D, neighbor_D, *_face_info, true);
+      face_velocity_interp_method, face_D, elem_D, neighbor_D, *_face_info, true);
 
-  // perform the pressure correction
+  // Perform the pressure correction. We don't use skewness-correction on these since
+  // it only influences the averaged vell gradients which cancel out in the correction
```
Comment (GiudGiud) [inline modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C:555]:
```suggestion
  // it only influences the averaged cell gradients which cancel out in the correction
```

File: framework/src/fvkernels/FVAdvection.C (hunk 8 above)
Comment (grmnptr) [reply framework/src/fvkernels/FVAdvection.C:27]: I had to add the skewness-corrected approach because the interpolation weights are different. We need to use the skewness-corrected weights to interpolate everything else to stay consistent in the scheme. Once I streamline the weights and interpolation approaches in the code, I'll make skewness correction a bool!

File: modules/navier_stokes/doc/content/modules/navier_stokes/insfv.md (hunk 11 above)
Comment (grmnptr) [reply modules/navier_stokes/doc/content/modules/navier_stokes/insfv.md:84]: The next commit will contain a specific example!

File: modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C (hunk 13 above)
Comment (grmnptr) [reply modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C:418]: Great idea! The latest commit checks the skewness-parameter for every direction separately.

File: framework/src/variables/MooseVariableFV.C (hunk 9 above)
Comment (grmnptr) [reply framework/src/variables/MooseVariableFV.C:1048]: This modification is a bugfix from the previous approach. In certain cases, if the code fails to compute the gradient with the two-term BD expansion, it falls back on this and without passing the skewness correction, it computes an uncorrected gradient with one-term BD expansion. I thought I can smuggle this fix into MOOSE with this PR.

File: modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/test.py (hunk 6 above)
Comment (lindsayad) [reply modules/navier_stokes/test/tests/finite_volume/ins/mms/skew-correction/test.py:28]: Haha of course. We can keep this in the back of our minds as something we'd like to understand at some point

File: framework/src/variables/MooseVariableFV.C (hunk 9 above)
Comment (lindsayad) [reply framework/src/variables/MooseVariableFV.C:1048]: In the legendary interpolator PR I've added a boolean to the `FaceArg` tuple for denoting skewness correction

File: framework/src/fvkernels/FVAdvection.C (hunk 8 above)
Comment (lindsayad) [reply framework/src/fvkernels/FVAdvection.C:27]: @GiudGiud are you satisfied with the response here?

File: framework/src/variables/MooseVariableFV.C (hunk 9 above)
Comment (lindsayad) [reply framework/src/variables/MooseVariableFV.C:1048]: So I think the change here is good. @GiudGiud please mark as resolved if you agree

**Review by lindsayad — APPROVED** [review]

**Comment by GiudGiud** [comment]
lgtm
